theseus/embodied/robot/forward_kinematics.py

```diff
@@ -0,0 +1,146 @@
+        poses = _forward_kinematics_helper(angles)
+        return tuple(poses[id] for id in link_ids)
+
+    def _jforward_kinematics_helper(
+        poses: List[Optional[torch.Tensor]],
+    ) -> torch.Tensor:
+        jposes = poses[0].new_zeros(poses[0].shape[0], 6, robot.dof)
+
+        for id in pose_ids[1:]:
+            link: Link = robot.links[id]
+            joint: Joint = link.parent
+            if joint.id >= robot.dof:
+                break
+            jposes[:, :, joint.id : joint.id + 1] = (
+                se3.adjoint(poses[link.id]) @ joint.axis
+            )
+
+        return jposes
+
+    def _jforward_kinematics_impl(angles: torch.Tensor):
+        poses = _forward_kinematics_helper(angles)
+        jposes = _jforward_kinematics_helper(poses)
+
+        rets = tuple(poses[id] for id in link_ids)
+        jacs: List[torch.Tensor] = []
+
+        for link_id in link_ids:
+            pose = poses[link_id]
+            jac = jposes.new_zeros(angles.shape[0], 6, robot.dof)
+            sel = robot.links[link_id].angle_ids
+            jac[:, :, sel] = se3.adjoint(se3.inverse(pose)) @ jposes[:, :, sel]
+            jacs.append(jac)
+
+        return jacs, rets
+
+    class ForwardKinematics(torch.autograd.Function):
+        @classmethod
+        def forward(cls, ctx, angles):
+            poses = _forward_kinematics_helper(angles)
+            ctx.poses = poses
+            rets = tuple(poses[id] for id in link_ids)
+            ctx.rets = rets
+            return rets
+
+        @classmethod
+        def backward(cls, ctx, *grad_outputs):
+            if not hasattr(ctx, "jposes"):
+                ctx.jposes: torch.Tensor = _jforward_kinematics_helper(ctx.poses)
+            rets: tuple(torch.Tensor) = ctx.rets
+            grad_pose = grad_outputs[0].new_zeros(
+                grad_outputs[0].shape[0], 6, robot.dof
+            )
+            grad_input = grad_outputs[0].new_zeros(grad_outputs[0].shape[0], robot.dof)
+
+            for link_id, ret, grad_output in zip(link_ids, rets, grad_outputs):
+                angle_ids = robot.links[link_id].angle_ids
+                temp = se3.project(
+                    torch.cat(
+                        (grad_output @ ret.transpose(1, 2), grad_output[:, :, 3:]),
+                        dim=-1,
+                    )
+                ).unsqueeze(-1)
+                grad_pose[:, :, angle_ids] += temp
+            grad_input[:, joint_ids] = (
+                ctx.jposes[:, :, joint_ids] * grad_pose[:, :, joint_ids]
+            ).sum(-2)
+
+            return grad_input
+
+    return (
+        ForwardKinematics,
+        _forward_kinematics_impl,
+        _jforward_kinematics_impl,
+        _forward_kinematics_helper,
+        _jforward_kinematics_helper,
+    )
+
+
+def get_forward_kinematics(robot: Robot, link_names: Optional[List[str]] = None):
+    ForwardKinematics, _, jforward_kinematics, _, _ = ForwardKinematicsFactory(
```

Comment:
Rename output as `ForwardKinematics` --> `fk_autograd_fn_cls`?

Reply:
Is there anything problematic about this suggestion? `CamelCase` is usually used for class definitions.

Reply:
I'm OK for either. I will address this comment after all the PRs have been merged since lots of functions use `ForwardKinematics`.